### PR TITLE
#880 Explicitly specifying componentModel="default" overrides the processor option…

### LINKED ISSUE #880: Specifying componentModel="default" in mapper has no effect if defaultComponentModel is configured
My assumption is that configuring defaultComponentModel literally overrides default component model.

So if I configure defaultComponentModel:
```xml
<options>
  <mapstruct.defaultComponentModel>spring</mapstruct.defaultComponentModel>
</options>
```

And then when I try to specify default as a component model for my mapper:
```java
@Mapper(componentModel = "default", uses=SomeOtherMapper.class)
public interface MyMapper {
```
it has no effect and the mapper is created as Spring bean and **SomeOtherMapper** is autowired in it instead of get using `Mappers.get(SomeOtherMapper.class)`.

I'm getting this in MapStruct 1.0.0.Final and 1.1.0.Beta2.

### PR DESCRIPTION
**Pull request by agudian**
This fixes #880 and is in sync with what's written in the Javadoc of `Mapper#componentModel`.

### CODE REVIEW

File: processor/src/main/java/org/mapstruct/ap/internal/util/MapperConfiguration.java [hunk 1]
```diff
@@ -170,12 +170,15 @@ private boolean isMapToDefaultOnMapperAndMappingConfigLevel() {
...
     public String componentModel() {
-        if ( mapperConfigPrism != null && mapperPrism.values.componentModel() == null ) {
-            return mapperConfigPrism.componentModel();
-        }
-        else {
+        if ( mapperPrism.values.componentModel() != null ) {
             return mapperPrism.componentModel();
         }
+
+        if ( mapperConfigPrism != null && mapperConfigPrism.values.componentModel() != null ) {
```
Comment (sjaakd) [inline processor/src/main/java/org/mapstruct/ap/internal/util/MapperConfiguration.java]: Hmm.. I wonder whether this same pattern is present for other configuration values...

**Comment by sjaakd** [comment]
looks good to me :+1: .. Please merge

File: processor/src/main/java/org/mapstruct/ap/internal/util/MapperConfiguration.java (hunk 1 above)
Comment (agudian) [inline processor/src/main/java/org/mapstruct/ap/internal/util/MapperConfiguration.java]: Good point, I just checked again and `unmappedTargetPolicy` looks suspicious.

Comment (agudian) [inline processor/src/main/java/org/mapstruct/ap/internal/util/MapperConfiguration.java]: Okay, `MapperConfig#unmappedTargetPolicy` is ignored if there was a value specified as processor option. Updating the PR...

**Comment by agudian** [comment]
I've replaced the whole thing with a different solution, to handle `componentModel` and `unmappedTargetPolicy` equally. Want to take another look, @sjaakd?

**Comment by sjaakd** [comment]
Of course. Will be tomorrow though!

File: processor/src/test/java/org/mapstruct/ap/test/bugs/_880/Poodle.java [hunk 2]
```diff
@@ -16,30 +16,20 @@
...
+package org.mapstruct.ap.test.bugs._880;
 
 /**
- * Helper class for dealing with {@link Options}.
- *
  * @author Andreas Gudian
+ *
  */
-public class OptionsHelper {
+public class Poodle {
```
Comment (sjaakd) [inline processor/src/test/java/org/mapstruct/ap/test/bugs/_880/Poodle.java]: So... It took me some time to put `Poodle` and `Core` together.... Its a literal translation of what means "The heart of the matter` or, on Dutch `De kern van de zaak`..  :)..

Looks a case of: "a strange duck in the pond"    :)

**Comment by sjaakd** [comment]
@agudian is ok for me.. 

I was just thinking, should `ReportingPolicy` not be a `ReportingPolicyPrism` (and be moved)?. It reflects an annotation value as well (to be consistent).

File: processor/src/test/java/org/mapstruct/ap/test/bugs/_880/Poodle.java (hunk 2 above)
Comment (agudian) [inline processor/src/test/java/org/mapstruct/ap/test/bugs/_880/Poodle.java]: It's a reference to Goethes 'Faust' in the English original... :wink:

**Comment by agudian** [comment]
> I was just thinking, should ReportingPolicy not be a ReportingPolicyPrism (and be moved)?. It reflects an annotation value as well (to be consistent).

Good catch! I did that and added two missing enum-prism tests.

**Comment by sjaakd** [comment]
@agudian, nice work.. Please go ahead and merge

**Pull request merged**